index.js

```diff
@@ -747,17 +747,25 @@
 
   //  Array$chainRec :: ((a -> c, b -> c, a) -> Array c, a) -> Array b
   function Array$chainRec(f, x) {
-    var $todo = [x];
-    var $done = [];
-    while ($todo.length > 0) {
-      var xs = f(iterationNext, iterationDone, $todo.shift());
-      var $more = [];
-      for (var idx = 0; idx < xs.length; idx += 1) {
-        (xs[idx].done ? $done : $more).push(xs[idx].value);
+    var result = [];
+    var nil = {};
+    var todo = {head: x, tail: nil};
+    while (todo !== nil) {
+      var more = nil;
+      f(iterationNext, iterationDone, todo.head).forEach(function(step) {
```

Comment:
Why switch to a `forEach` here? 

As much as I _hate_ for loops in application code, in a library I _think_ it's faster than a `forEach`

Reply:
Now that Aldwin has added benchmarks (:clap:) we can see. I'll compare the two options.

Reply:
Using a `for` loop here is significantly faster on my computer. Thanks for the suggestion, Brad!